// amqplib-mocks/lib/connection.js
"use strict";

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const _ = require("lodash");
const sinon = require("sinon");
const Channel = require("./channel");

class Connection {
	constructor(onClose) {
		this.queues = {};
		this.exchanges = {};
		this.on = sinon.stub();

		this.createChannel = sinon.stub().callsFake(() => {
			return this.channel = new Channel(this);
		});

		this.createConfirmChannel = sinon.stub().callsFake(() => {
			return this.channel = new Channel(this);
		});

		this.close = sinon.stub().callsFake(() => {
			onClose();
			this.on.withArgs("close").yield();
		});
	}

	// Test Helpers

	get currentChannel() {
		return this.channel;
	}

	get trackedMessages() {
		return this.channel.trackedMessages;
	}

	sendUntracked(queueName, content, properties) {
		var _this = this;

		return _asyncToGenerator(function* () {
			const channel = new Channel(_this);
			yield channel.sendToQueue(queueName, content, properties);
		})();
	}

	publishUntracked(exchange, routingKey, content, properties) {
		var _this2 = this;

		return _asyncToGenerator(function* () {
			const channel = new Channel(_this2);
			yield channel.publish(exchange, routingKey, content, properties);
		})();
	}

	getPublished({ filter = _.stubTrue, bodyTransform = _.identity }) {
		return this.channel.trackedMessages.filter(msg => filter(msg)).map(msg => Object.assign(msg, { body: bodyTransform(msg.content) }));
	}
}

module.exports = Connection;
